jso/apis/src/main/java/org/teavm/jso/core/JSPromise.java

```diff
@@ -0,0 +1,103 @@
+/*
+ *  Copyright 2023 Bernd Busse.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.teavm.jso.core;
+
+import org.teavm.interop.NoSideEffects;
+import org.teavm.jso.JSBody;
+import org.teavm.jso.JSByRef;
+import org.teavm.jso.JSFunctor;
+import org.teavm.jso.JSMethod;
+import org.teavm.jso.JSObject;
+
+import org.teavm.jso.util.function.JSConsumer;
+import org.teavm.jso.util.function.JSFunction;
+import org.teavm.jso.util.function.JSSupplier;
+
+/**
+ * Interface for interacting with JavaScript <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise">{@code Promise}s</a>.
+ *
+ * @param <T> The type this promise returns when resolving successfully.
+ * @param <E> The type this promise returns when rejecting.
+ */
+public abstract class JSPromise<T, E> implements JSObject {
```

Comment:
I think type parameter `E` is unnecessary. In TypeScript, `Promise` interface only declares type parameter for value type.

Reply:
Alright, I'll leave the `reject` case as a generic `Object` then.
I seem to remember a convention of rejecting with `Error` instances. But I don't think we can currently create `Error` instances directly.